### ISSUE TITLE
RF06: Unnecessary Quoted Identifier error when column contains special character in identifier

### ISSUE THREAD

**Issue by rwork0** [sqlfluff/sqlfluff#6717]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When the column name is quoted as it contains a special character and the name contains no lower case characters, the RF06 rule marks the quotes as unnecessary.

Example query:
```sql
SELECT [ABC#] AS [abc_count]
FROM foo

``` 

### Expected Behaviour

The rule should not fail.

### Observed Behaviour

== [test.sql] FAIL
L:   1 | P:   8 | RF06 | Unnecessary quoted identifier [ABC#].
                       | [references.quoting]

### How to reproduce

test.sql>
SELECT [ABC#] AS [abc_count]
FROM foo

`sqlfluff lint test.sql --dialect tsql`

### Dialect

tsql

### Version

sqlfluff, version 3.3.1
Python 3.11.7

### Configuration

N/A

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by peterbud** [comment]
Hi @rwork0, thanks for submitting this issue.

According to [SQL Server Documentation](https://learn.microsoft.com/en-us/sql/relational-databases/databases/database-identifiers?view=sql-server-ver17), the `ABC#` identifier is perfectly fine and would NOT require quotes. I have tested and SQL Server is happily executing 

```sql
SELECT ABC# AS [abc_count]
FROM foo
```

So the `RF06` rule is right in this case.

**Comment by peterbud** [comment]
Closing for now
